tracing: add span for get cacher

vrutkovs opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Create a tracing span for `Cacher`'s `Get` so that it would be easy to see if GET request was cached

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Commit:
tracing: add span for cacher.Get

Also updates tracing integration tests for cacher.GetList

vrutkovs commented:
/cc @dashpole

seans3 commented:
/remove-sig api-machinery
/sig instrumentation

seans3 commented:
/cc @dashpole

dashpole reviewed (commented):
Can we add checks for this span to the [integration test](https://github.com/kubernetes/kubernetes/blob/master/test/integration/apiserver/tracing/tracing_test.go)?

We should also consider updating the span names and attributes to follow otel conventions.  Names don't usually contain spaces, for example, and attributes should have a prefix (probably k8s.foo)

vrutkovs commented:
/kind bug

>We should also consider updating the span names and attributes to follow otel conventions. Names don't usually contain spaces, for example

Ah, sure, I also renamed "cacher list" to "CacherList". As for attributes - I think it needs to be separately discussed and implemented ( exact prefix, `audit.id` instead of `audit-id` etc.)

dashpole reviewed (approved):
/lgtm

dgrisonnet commented:
/triage accepted
/ok-to-test

vrutkovs commented:
/cc wojtek-t apelisse

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:
```diff
@@ -107,7 +107,7 @@ func GetResource(r rest.Getter, scope *RequestScope) http.HandlerFunc {
 				}
 			}
-			tracing.SpanFromContext(ctx).AddEvent("About to Get from storage")
+			tracing.SpanFromContext(ctx).AddEvent("About to Get from cacher")
```
I don't agree with this change - at this point we don't know if it will be server from cacher or from etcd.

Please revert here and below.

vrutkovs replied:
It would first hit cacher at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L695 and only afterwards we'll get the event if it hits actual storage, am I understanding this correctly?

wojtek-t replied:
if cacher is enabled (e.g. it's not enabled for events by default)

also - cacher is an implementation of storage, so let's not change that.

vrutkovs replied:
oh, so "storage" in this context is "storage implementation, which starts with cacher" and not "underlying storage (read: etcd)"? Its a bit confusing that the next span would be "cacher.GetList" though. 

How about:
* `About to Get from cacher` here
* cacher.Get or cacher.List is called
* if cache is uninitialized or missed the event `About to Get from storage implementation` is emitted?

wojtek-t replied:
Cacher doesn't have to be there - and by default it's not enabled for events.
We shouldn't mention cacher here.

vrutkovs replied:
Updated, PTAL

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -856,7 +865,7 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	}
 
-	ctx, span := tracing.Start(ctx, "cacher list",
+	ctx, span := tracing.Start(ctx, "CacherList",
```
Why do you change that? I think it's misleading.

I could by `Cacher List`, but there isn't anything like "CacherList"...

vrutkovs replied:
According to https://github.com/open-telemetry/opentelemetry-specification/blob/v1.35.0/specification/trace/api.md#span it should represent a function, so spaces in the name are undesirable.

I don't mind renaming it to something more identifiable though

dashpole replied:
`cacher.list` might be better?

vrutkovs replied:
`cacher.GetList` sounds better to me

vrutkovs replied:
Renamed

vrutkovs commented:
/test pull-kubernetes-unit
